src/tools/rbd_mirror/ImageReplayer.cc

```diff
@@ -176,6 +176,225 @@ void ImageReplayer<I>::BootstrapProgressContext::update_progress(
 }
 
 template <typename I>
+void ImageReplayer<I>::ReplayStatusMaker::send(Context *on_finish) {
+  dout(20) << dendl;
+
+  if (m_on_finish) {
+    dout(20) << "previous request is still in progress, ignoring" << dendl;
+    on_finish->complete(-EAGAIN);
+    return;
+  }
+
+  m_on_finish = on_finish;
+  get_registered_clients();
+}
+
+template <typename I>
+void ImageReplayer<I>::ReplayStatusMaker::get_registered_clients() {
```

Comment:
@dillaman Looking at it again, actually querying the registered clients instead of using the cached ones gives the same results and does not resolve possible negative entries_behind_master:
```
2016-04-26 09:15:12.180129 7f5e2effd700 20 rbd-mirror: ImageReplayer[3/103b74b0dc51]::ReplayStatusMaker::make_description: m_master_position=[object_number=3, tag_tid=2, entry_tid=3], m_mirror_position=[object_number=3, tag_tid=2, entry_tid=3], m_entries_behind_master=0
2016-04-26 09:15:42.181220 7f5e2effd700 20 rbd-mirror: ImageReplayer[3/103b74b0dc51]::ReplayStatusMaker::make_description: m_master_position=[object_number=3, tag_tid=3, entry_tid=859], m_mirror_position=[object_number=3, tag_tid=3, entry_tid=1471], m_entries_behind_master=-612
2016-04-26 09:16:12.182310 7f5e2effd700 20 rbd-mirror: ImageReplayer[3/103b74b0dc51]::ReplayStatusMaker::make_description: m_master_position=[object_number=2, tag_tid=3, entry_tid=2702], m_mirror_position=[object_number=3, tag_tid=3, entry_tid=3327], m_entries_behind_master=-625
2016-04-26 09:16:30.526062 7f5e2effd700 20 rbd-mirror: ImageReplayer[3/103b74b0dc51]::ReplayStatusMaker::make_description: m_master_position=[object_number=3, tag_tid=3, entry_tid=4319], m_mirror_position=[object_number=3, tag_tid=3, entry_tid=4319], m_entries_behind_master=0
2016-04-26 09:17:00.527246 7f5e2effd700 20 rbd-mirror: ImageReplayer[3/103b74b0dc51]::ReplayStatusMaker::make_description: m_master_position=[object_number=2, tag_tid=4, entry_tid=790], m_mirror_position=[object_number=3, tag_tid=4, entry_tid=1503], m_entries_behind_master=-713
2016-04-26 09:17:30.528257 7f5e2effd700 20 rbd-mirror: ImageReplayer[3/103b74b0dc51]::ReplayStatusMaker::make_description: m_master_position=[object_number=2, tag_tid=4, entry_tid=2734], m_mirror_position=[object_number=3, tag_tid=4, entry_tid=3391], m_entries_behind_master=-657
2016-04-26 09:17:48.072964 7f5e2effd700 20 rbd-mirror: ImageReplayer[3/103b74b0dc51]::ReplayStatusMaker::make_description: m_master_position=[object_number=3, tag_tid=4, entry_tid=4319], m_mirror_position=[object_number=3, tag_tid=4, entry_tid=4319], m_entries_behind_master=0
2016-04-26 09:18:18.074030 7f5e2effd700 20 rbd-mirror: ImageReplayer[3/103b74b0dc51]::ReplayStatusMaker::make_description: m_master_position=[object_number=3, tag_tid=5, entry_tid=855], m_mirror_position=[object_number=3, tag_tid=5, entry_tid=1663], m_entries_behind_master=-808
2016-04-26 09:18:48.086746 7f5e2effd700 20 rbd-mirror: ImageReplayer[3/103b74b0dc51]::ReplayStatusMaker::make_description: m_master_position=[object_number=3, tag_tid=5, entry_tid=3003], m_mirror_position=[object_number=3, tag_tid=5, entry_tid=3359], m_entries_behind_master=-356
2016-04-26 09:19:18.088519 7f5e2effd700 20 rbd-mirror: ImageReplayer[3/103b74b0dc51]::ReplayStatusMaker::make_description: m_master_position=[object_number=3, tag_tid=5, entry_tid=4319], m_mirror_position=[object_number=3, tag_tid=5, entry_tid=4319], m_entries_behind_master=0
2016-04-26 09:19:22.354803 7f5e2effd700 20 rbd-mirror: ImageReplayer[3/103b74b0dc51]::ReplayStatusMaker::make_description: m_master_position=[object_number=3, tag_tid=5, entry_tid=4319], m_mirror_position=[object_number=3, tag_tid=5, entry_tid=4319], m_entries_behind_master=0
2016-04-26 09:19:52.355642 7f5e2effd700 20 rbd-mirror: ImageReplayer[3/103b74b0dc51]::ReplayStatusMaker::make_description: m_master_position=[object_number=0, tag_tid=6, entry_tid=1204], m_mirror_position=[object_number=3, tag_tid=6, entry_tid=1727], m_entries_behind_master=-523
2016-04-26 09:20:25.273524 7f5e2effd700 20 rbd-mirror: ImageReplayer[3/103b74b0dc51]::ReplayStatusMaker::make_description: m_master_position=[object_number=7, tag_tid=6, entry_tid=4319], m_mirror_position=[object_number=0, tag_tid=6, entry_tid=2140], m_entries_behind_master=2179
2016-04-26 09:20:55.274942 7f5e2effd700 20 rbd-mirror: ImageReplayer[3/103b74b0dc51]::ReplayStatusMaker::make_description: m_master_position=[object_number=7, tag_tid=6, entry_tid=4319], m_mirror_position=[object_number=4, tag_tid=6, entry_tid=4316], m_entries_behind_master=3
```
 I think this is because the mirror commits its position after every entry apply, while the master client may delay.

I will use the cached clients, though there is no idea if negative entries_behind_master can be fixed/improved somehow.